keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,389 @@
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - name: Zookeeper Configuration File
+        key: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+            name: "{{ .InstanceName }}-zookeeper"
+      - name: DNS information for running pod
+        key: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          objectRef:
+            kind: Pod
+            name: "{{ .InstanceName }}-zookeeper-0"
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
+of the `diagnostics.bundle` key's presence. Note, moving to a graph-based engine
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Steps in a bundle run serially. To prevent the KUDO controller manager from
+crashing, the collector process runs in another pod as a job. **TODO**:
+Bundle collection, CRD, do we take a Velero-style approach? Where are files
+stored? Might be time to introduce a KUDO-specific Minio instance.
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.filterSecrets` key.
```

Comment:
This is the only time `filterSecrets` is mentioned. Is it different than `diagnostics.bundle.filters`?

Reply:
@gerred could you please confirm that filterSecrets was indeed meant as a specific field?